client/src/main/java/org/apache/celeborn/client/write/InFlightTracker.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.celeborn.client.write;
+
+import java.io.IOException;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.celeborn.common.CelebornConf;
+
+public class InFlightTracker {
+  private static final Logger logger = LoggerFactory.getLogger(InFlightTracker.class);
+
+  private final long timeoutMs;
+  private final long delta;
+  private final int maxInFlight;
+
+  private final PushState pushState;
+
+  public final AtomicInteger batchId = new AtomicInteger();
+  private final ConcurrentHashMap<String, Set<Integer>> batchIdPerAddressPair =
+      new ConcurrentHashMap<>();
+
+  public InFlightTracker(CelebornConf conf, PushState pushState) {
+    this.timeoutMs = conf.pushLimitInFlightTimeoutMs();
+    this.delta = conf.pushLimitInFlightSleepDeltaMs();
+    this.maxInFlight = conf.pushMaxReqsInFlight();
+    this.pushState = pushState;
+  }
+
+  public void addFlightBatches(int batchId, String hostAndPushPort) {
```

Comment:
Should we add sync for such method about `batchIdPerAddressPair `? @waitinfuture